File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequency = {}
        for num in nums:
            if num in frequency:
                frequency[num] += 1
            else:
                frequency[num] = 1

        frequency_list = [(num, freq) for num, freq in frequency.items()]

        frequency_list.sort(key=lambda x: x[1], reverse=True)

        top_k = [num for num, freq in frequency_list[:k]]

        return top_k
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        //num : frequency
        HashMap<Integer, Integer> freq = new HashMap<>();
        
        for(int i=0; i<nums.length; i++){
            freq.put(nums[i], freq.getOrDefault(nums[i], 0) + 1);
        }
        
        PriorityQueue<Map.Entry<Integer, Integer>> minHeap = new PriorityQueue<>(
            (a, b) -> a.getValue() - b.getValue());
        
        for (Map.Entry<Integer, Integer> entry : freq.entrySet()) {
            minHeap.add(entry);
            if (minHeap.size() > k) {
                minHeap.poll(); 
            }
        }
        
        int[] res = new int[k];
        
        for (int i = k - 1; i >= 0; i--) { 
            res[i] = minHeap.poll().getKey();
        }

        return res;
    }
}